File: channel/lib/api/media.ts
import { apiClient, getApiUrl, ApiResponse } from "./client"

export type MediaType = "image" | "video" | "audio" | "document" | "sticker"

export interface MediaFile {
  id: string
  sessionId: string
  msgId: string
  chatJid: string
  mediaType: MediaType
  mimeType: string
  fileSize: number
  fileName?: string
  storageUrl?: string
  storageKey?: string
  thumbnailUrl?: string
  downloaded: boolean
  downloadError?: string
  createdAt: string
}

export interface MediaListParams {
  limit?: number
  offset?: number
  mediaType?: MediaType
  chatJid?: string
}

export async function getMediaList(
  sessionId: string,
  params?: MediaListParams
): Promise<ApiResponse<MediaFile[]>> {
  const searchParams = new URLSearchParams()
  if (params?.limit) searchParams.set("limit", params.limit.toString())
  if (params?.offset) searchParams.set("offset", params.offset.toString())
  if (params?.mediaType) searchParams.set("mediaType", params.mediaType)
  if (params?.chatJid) searchParams.set("chatJid", params.chatJid)
  const query = searchParams.toString()
  return apiClient<MediaFile[]>(`/sessions/${sessionId}/media/list${query ? `?${query}` : ""}`)
}

export async function getPendingMedia(
  sessionId: string
): Promise<ApiResponse<MediaFile[]>> {
  return apiClient<MediaFile[]>(`/sessions/${sessionId}/media/pending`)
}

export async function processMedia(
  sessionId: string,
  mediaIds?: string[]
): Promise<ApiResponse<void>> {
  return apiClient<void>(`/sessions/${sessionId}/media/process`, {
    method: "POST",
    body: JSON.stringify({ ids: mediaIds }),
  })
}

export async function retryMediaDownload(
  sessionId: string,
  mediaId: string
): Promise<ApiResponse<void>> {
  return apiClient<void>(`/sessions/${sessionId}/media/retry`, {
    method: "POST",
    body: JSON.stringify({ id: mediaId }),
  })
}

export function getMediaStreamUrl(sessionId: string, mediaId: string): string {
  return getApiUrl(`/public/${sessionId}/media/stream?id=${mediaId}`)
}

export function getMediaDownloadUrl(sessionId: string, mediaId: string): string {
  return getApiUrl(`/sessions/${sessionId}/media/download?id=${mediaId}`)
}

export function formatFileSize(bytes: number): string {
  if (bytes === 0) return "0 Bytes"
  const k = 1024
  const sizes = ["Bytes", "KB", "MB", "GB"]
  const i = Math.floor(Math.log(bytes) / Math.log(k))
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + " " + sizes[i]
}

export function getMediaIcon(mediaType: MediaType): string {
  switch (mediaType) {
    case "image":
      return "Image"
    case "video":
      return "Video"
    case "audio":
      return "Music"
    case "document":
      return "FileText"
    case "sticker":
      return "Sticker"
    default:
      return "File"
  }
}

File: api/internal/model/media.go
package model

import "time"

type MediaType string

const (
	MediaTypeImage    MediaType = "image"
	MediaTypeVideo    MediaType = "video"
	MediaTypeAudio    MediaType = "audio"
	MediaTypeDocument MediaType = "document"
	MediaTypeSticker  MediaType = "sticker"
)

type Media struct {
	ID        string `json:"id"`
	SessionID string `json:"sessionId"`
	MsgID     string `json:"msgId"`

	MediaType string `json:"mediaType"`
	MimeType  string `json:"mimeType,omitempty"`
	FileSize  int64  `json:"fileSize,omitempty"`
	FileName  string `json:"fileName,omitempty"`

	WADirectPath        string `json:"waDirectPath,omitempty"`
	WAMediaKey          []byte `json:"waMediaKey,omitempty"`
	WAFileSHA256        []byte `json:"waFileSha256,omitempty"`
	WAFileEncSHA256     []byte `json:"waFileEncSha256,omitempty"`
	WAMediaKeyTimestamp int64  `json:"waMediaKeyTimestamp,omitempty"`

	Width    int `json:"width,omitempty"`
	Height   int `json:"height,omitempty"`
	Duration int `json:"duration,omitempty"`

	StorageKey string     `json:"storageKey,omitempty"`
	StorageURL string     `json:"storageUrl,omitempty"`
	StoredAt   *time.Time `json:"storedAt,omitempty"`

	ThumbnailKey string `json:"thumbnailKey,omitempty"`
	ThumbnailURL string `json:"thumbnailUrl,omitempty"`

	Downloaded       bool   `json:"downloaded"`
	DownloadError    string `json:"downloadError,omitempty"`
	DownloadAttempts int    `json:"downloadAttempts"`

	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`

	ChatJID  string `json:"chatJid,omitempty"`
	FromMe   bool   `json:"fromMe,omitempty"`
	Caption  string `json:"caption,omitempty"`
	PushName string `json:"pushName,omitempty"`
}

func (m *Media) CanDownload() bool {
	return m.WADirectPath != "" && len(m.WAMediaKey) > 0
}

func (m *Media) NeedsDownload() bool {
	return !m.Downloaded && m.CanDownload() && m.DownloadAttempts < 3
}
